### ISSUE TITLE
MPIVecEnv

### ISSUE THREAD

**Issue by sgillen** [Stable-Baselines-Team/stable-baselines3-contrib#45]
Hello, 

I was trying to find a way to make the ARS implementation I was working on [in this pr](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/42) faster. My first thought was a drop in replacement for SubprocVecEnv that uses mpi4py instead. I implemented a first pass [here](https://github.com/sgillen/stable-baselines3-contrib/blob/feat/mpi_vec_env/sb3_contrib/common/envs/mpi_vec_env.py). It is quick and dirty, but still a working proof of concept to see if there is any performance to be gained here. 

I am seeing modest speedups in rollout collection time. For pendulum-v0 with 10 environments I am finding it 4-5x faster than Dummy and Subproc. For HumanoidBulletEnv-v0 with 10 environments I am finding it 8x faster than Dummy and 2x Faster than SubProc. It might be possible to squeeze more performance out of it but probably this is 80% of what can be achieved using this approach. 

This is just for rollout collections, any actual speedup to algorithms using this vec env are going to be smaller, but for on policy algorithms probably still significant. 

I wanted to ask if this or something like it had been considered, IIRC mpi4py was a big headache to support, but perhaps by confining that dependency to contrib/ most of the headache will disappear. Can also look at for example [torch distributed](https://pytorch.org/docs/stable/distributed.html), but I think that will cause a similar number of headaches for most likely less speed. 

This is another thing I would interested in contributing (over the following weeks ...). But again, only if there is interest.

**Comment by araffin** [comment]
Hello,
it looks interesting, do you have a minimal code example on how it works?
the synchronization is done after each step?

In fact, I have the plan to have a complete PPO MPI version at some point in SB3 contrib (#11 ) where each core also compute gradients.

**Comment by sgillen** [comment]
Yes, the examples I made for testing live [here](https://github.com/sgillen/stable-baselines3-contrib/tree/feat/mpi_vec_env/sb3_contrib/common/envs). The synchronization happens at each steps, each call to G/gather is blocking. 

I figure the proposed vec env could be used by any existing sb3 algorithm for a modest speedup, and maybe by future "MPI aware" algorithms for bigger gains.

**Comment by araffin** [comment]
> Yes, the examples I made for testing live here. 

thanks =) but I'm afraid this won't work with the current SB3 implementation...

**Comment by sgillen** [comment]
OK, I'm not seeing why that is though?
